src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt

```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
+        configureTalonsForVelocityControl()
+        currentState = DriveControlState.VELOCITY_SETPOINT
+        updateVelocitySetpoint(left_inches_per_sec, right_inches_per_sec)
+    }
+
+    @Synchronized
+    private fun updateVelocitySetpoint(leftInchesPerSec: Double, rightInchesPerSec: Double){
+        if(usesTalonVelocityControl(currentState)){
+            val maxDesired: Double=Math.max(Math.abs(leftInchesPerSec), Math.abs(rightInchesPerSec))
+            var scale: Double;
+            if (maxDesired > Constants.Drive.HIGH_GEAR_MAX_SETPOINT){
+                scale=Constants.Drive.HIGH_GEAR_MAX_SETPOINT/ maxDesired
+            }
+            else{
+                scale=1.0
+            }
+            leftMasterSRX.set(ControlMode.Velocity,inchesPerSecondToRpm(leftInchesPerSec*scale))
+            rightMasterSRX.set(ControlMode.Velocity,inchesPerSecondToRpm(rightInchesPerSec*scale))
+        }
+        else {
+            print("Incorrect Velocity Control Mode")
+            leftMasterSRX.set(ControlMode.Velocity,0.0)
+            rightMasterSRX.set(ControlMode.Velocity,0.0)
+        }
+
+    }
+
+    @Synchronized
+    private fun updatePositionSetpoint(leftPositionInches: Double, rightPositionInches: Double){
+        if (usesTalonPositionControl(currentState)){
+            leftMasterSRX.set(ControlMode.MotionMagic,leftPositionInches)
+            leftMasterSRX.set(ControlMode.MotionMagic,leftPositionInches)
+        }
+        else {
+            print("Bad position control state")
+            leftMasterSRX.set(ControlMode.MotionMagic,0.0)
+            rightMasterSRX.set(ControlMode.MotionMagic,0.0)
+        }
+    }
+
+    private fun configureTalonsForVelocityControl() { //should further review cause im bad
+        if (!usesTalonVelocityControl(currentState)) {
+            // We entered a velocity control state.
+            leftMasterSRX.set(ControlMode.Velocity,0.0) //velocity  output value is in position change / 100ms
+            leftMasterSRX.configNominalOutputReverse(Constants.Velocity.driveHighGearNominalOutput, 10)
+            leftMasterSRX.selectProfileSlot(Constants.Velocity.highGearVelocityControlSlot, 0)
+            leftMasterSRX.configPeakOutputForward(Constants.Velocity.driveHighGearNominalOutput, 10)
+            rightMasterSRX.set(ControlMode.Velocity,0.0) //velocity  output value is in position change / 100ms
+            rightMasterSRX.configNominalOutputReverse(Constants.Velocity.driveHighGearNominalOutput, 10)
+            rightMasterSRX.selectProfileSlot(Constants.Velocity.highGearVelocityControlSlot, 0)
+            rightMasterSRX.configPeakOutputForward(Constants.Velocity.driveHighGearNominalOutput, 10)
+            setBrakeMode(NeutralMode.Brake)
+        }
+    }
+     private fun configureTalonsforPositionControl(){
+         if (!usesTalonPositionControl(currentState)) {
+             // We entered a position control state.
+             leftMasterSRX.set(ControlMode.MotionMagic,0.0);
```

Comment:
Is this a delta value, or does it try to set your position to zero from where it started? If the second one, this absolutely should not be zero unless you reset the encoders.